feat(common): add automatic srcset feature to ngOptimizedImage

atcastle opened the pull request:
This PR adds a feature for automatic srcset generation to the NgOptimizedImage directive. Srcsets are generated as density selectors (eg "1x", "2x") if there is no `sizes` prop provided, or with responsive srcsets with a range of pixel values if a `sizes` prop exists. These responsive breakpoints are user-configurable using a new provider called ImageConfig.

The automatic srcset can be disabled with a new `unoptimized` attribute, or via the ImageConfig provider.

Note: This PR only contains unit testing of this feature. End-to-end tests will be added shortly. Opening this now to begin discussion with @pkozlowski-opensource and @AndrewKushnir.

// CC: @kara 

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Feature


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

kara commented on aio/content/guide/image-directive.md:
```diff
@@ -89,24 +89,64 @@ You can typically fix this by adding `height: auto` or `width: auto` to your ima
 
 ### Handling `srcset` attributes
```
```suggestion
### Requesting images at the right size
```

kara commented on aio/content/guide/image-directive.md:
```diff
@@ -89,24 +89,64 @@ You can typically fix this by adding `height: auto` or `width: auto` to your ima
 ### Handling `srcset` attributes
 
-If your `<img>` tag defines a `srcset` attribute, replace it with `ngSrcset`.
+'NgOptimizedImage' generates the [`srcset` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset) for you automatically, based on the configured image loader. The generated `srcset` is dependent on the [`sizes` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes) passed to the directive. If `sizes` is omitted, it is assumed that your im ...
```
Can we add a few lines to explain what srcset/sizes mean? As I was trying to provide some suggestions below, I'm wondering if we're asking for folks to learn too much about how "sizes" works (and it's kind of complicated!). If we had a higher-level API (like `responsive="100vw"`), it would be easier to learn and we could fall through to "sizes".

Anyway, here's a stab at adding more info about srcset/sizes, but I think we'd really need to add more context:

```suggestion
Defining a [`srcset` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset) ensures that the browser requests an image at the right size for your user's viewport, so it doesn't waste time downloading an image that's too large. 'NgOptimizedImage' generates an appropriate `srcset` for the image, based on the presence and value of the [`sizes` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes) on the image tag.

#### Fixed size images

If your image should be "fixed" in size  (i.e. the same size across devices), there is no need to set a `sizes` attribute. A `srcset` can be generated automatically from the image's width and height attributes with no further input required. 

Example srcset generated: `<img ... srcset="image-400w.jpg 1x, image-800w.jpg 2x">`

#### Responsive images

If your image should be responsive (i.e. grow and shrink according to viewport size), then you will need to define a [`sizes` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes) to generate the `srcset`.

If you haven't used `sizes` before, a good place to start is to set it based on viewport width. For example, if your CSS causes the image to fill 100% of viewport width, set `sizes` to `100vw` and the browser will select the image in the `srcset` that is closest to 100% vw (after accounting for pixel density). If your image is only likely to take up half the screen (ex: in a sidebar), set `sizes` to `50vw` to ensure the browser selects a smaller image. And so on.

By default, these sizes are:
```

kara commented on aio/content/guide/image-directive.md:
```diff
@@ -89,24 +89,64 @@ You can typically fix this by adding `height: auto` or `width: auto` to your ima
-If the `ngSrcset` attribute is present, `NgOptimizedImage` generates and sets the [`srcset` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset) using the configured image loader. Do not include image file names in `ngSrcset` - the directive infers this information from `ngSrc`. The directive supports both width descriptors (e.g. `100w`) and density descriptors (e. ...
-
-You can also use `ngSrcset` with the standard image [`sizes` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes).
+If the `ngSrcset` attribute is present, `NgOptimizedImage` generates and sets the  using the configured image loader. Do not include image file names in `ngSrcset` - the directive infers this information from `ngSrc`. The directive supports both width descriptors (e.g. `100w`) and density descriptors (e.g. `1x`) are supported.
```
How do `ngSrcset` and the generated `srcset` intersect? Is there any reason to keep `ngSrcset` around if you can define custom breakpoints anyway?

atcastle replied:
Right now if you use `ngSrcset` the directive will just use that value, and won't use the automatic srcset. I'm not sure if we need to keep providing the ngSrcset convenience API or not. It could be useful for some people, but it also does complicate the API

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -370,6 +422,51 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+    const {viewportBreakpoints, subViewportBreakpoints} = this.config;
+
+    let breakpoints: number[];
+    if (this.sizes?.trim() === '100vw') {
```
Add a comment that explains why we only need viewportBreakpoints here?

atcastle replied:
Done

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -370,6 +422,51 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+      // the smallest percent size in the `sizes` prop, multiplying it by the smallest
+      // size in the viewport prop, and removing anything smaller from the subviewport
+      // breakpoints.
+      const vwRegExp = /(^|\s)(1?\d?\d)vw/g;
```
How necessary do we feel this trimming is? It seems like more computation to do for every image, and I'm not sure it's worth saving a few bytes in the document.

kara replied:
Also what happens if the "sizes" attribute contains a fancy query (e.g. "(max: 500px) 400w" type thing)? Do we want to support that? I guess we'd have to throw it contained a value that wasn't in our breakpoint list

atcastle replied:
Yeah, I'd be ok with removing this logic. Especially in a CSR scenario, there's not that much benefit to trimming those extra URLs. Though, it could make some sort of a difference on a page with a lot of images doing SSR. A few of the smallest breakpoints (16, 32) tend get trimmed almost all the time.

As for the mixed queries--I don't think that's super common, and I also think they're a pretty good case for where someone would want to shut off the automated srcset entirely. If you're using sizes with that level of specificity, you probably also want to manually define your srcset.

kara replied:
Let's remove this trimming logic for now? We can discuss and add it in a future PR if we think it's necessary

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -370,6 +422,51 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+  }
+
+  private getFixedSrcset(): string {
+    const multipliers = [1, 2];
```
Let's move the supported density descriptors into a constant at the top of the file?

atcastle replied:
Whoops, meant to do just that. Good catch

kara reviewed (commented):
Thanks for the PR! Great basis for discussion on srcset design - let's continue in the chat?

atcastle commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +61,35 @@ const ASPECT_RATIO_TOLERANCE = .1;
 const OVERSIZED_IMAGE_TOLERANCE = 1000;
 
+/**
+ * Represents an image loader function. Image loader functions are used by the
```
Done

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -370,6 +430,52 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+      let percentSizeMatch;
+      const sizeRatios = [];
+      while (percentSizeMatch = vwRegExp.exec(this.sizes!)) {
+        sizeRatios.push(parseInt(percentSizeMatch[2]) * 0.01);
```
```suggestion
        sizeRatios.push(parseInt(percentSizeMatch[2], 10) * 0.01);
```
Could you please also make `0.01` a const (so that we can document what it represents)?

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -399,6 +505,19 @@ function inputToBoolean(value: unknown): boolean {
+ * Sorts provided config breakpoints and uses defaults.
+ */
+function processConfig(config: ImageConfig): ImageConfig {
+  let sortedBreakpoints: {subViewportBreakpoints?: number[], viewportBreakpoints?: number[]} = {};
+  if (config.subViewportBreakpoints) {
```
nit: let's do object destructing here to same some bytes (after minificattion all property names remain the same, when const names are minified):
```suggestion
  const {viewportBreakpoints, subViewportBreakpoints} = config;
  let sortedBreakpoints: {subViewportBreakpoints?: number[], viewportBreakpoints?: number[]} = {};
  if (subViewportBreakpoints) {
```

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -430,6 +549,20 @@ function assertNoConflictingSrcset(dir: NgOptimizedImage) {
   }
 }
 
+function assertNoOverlappingBreakpoints(dir: NgOptimizedImage, config: ImageConfig) {
```
Could you please add a JSDoc to this comment to  document what it verifies?

atcastle replied:
Removed this error case entirely since there's now only one breakpoint array.

AndrewKushnir commented on packages/common/src/errors.ts:
```diff
@@ -32,4 +32,5 @@ export const enum RuntimeErrorCode {
   UNEXPECTED_DEV_MODE_CHECK_IN_PROD_MODE = 2958,
   INVALID_LOADER_ARGUMENTS = 2959,
   OVERSIZED_IMAGE = 2960,
+  OVERLAPPING_BREAKPOINTS = 2961
```
We can probably add a comma at the end here as well (similar to last error code on the current list):
```suggestion
  OVERLAPPING_BREAKPOINTS = 2961,
```

atcastle replied:
Removed this error case entirely since there's now only one breakpoint array.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -370,6 +430,52 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+      }
+      if (sizeRatios.length) {
+        let smallestRatio = Math.min(...sizeRatios);
+        breakpoints = breakpoints.filter(bp => bp >= viewportBreakpoints![0] * smallestRatio);
+      }
+    }
+
+    const finalSrcs = breakpoints.map(
+        breakpoint => `${this.imageLoader({src: this.ngSrc, width: breakpoint})} ${breakpoint}w`);
```
nit: for consistency, we can either call the argument `bp` or `breakpoint` in both places.

atcastle replied:
Done. Used "bp" in both

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -370,6 +430,52 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+      // breakpoints.
+      const vwRegExp = /(^|\s)(1?\d?\d)vw/g;
+      let percentSizeMatch;
+      const sizeRatios = [];
+      while (percentSizeMatch = vwRegExp.exec(this.sizes!)) {
+        sizeRatios.push(parseInt(percentSizeMatch[2]) * 0.01);
```
I think we can avoid an intermediate array here without affecting the reasability:
```suggestion
      let smallestRatio: number|null = null;
      let percentSizeMatch;
      while (percentSizeMatch = vwRegExp.exec(this.sizes!)) {
        smallestRatio = Math.min(smallestRatio, parseInt(percentSizeMatch[2]) * 0.01);
      }
      if (smallestRatio !== null) { ... }
```

atcastle replied:
Looks good! Updated

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -430,6 +549,20 @@ function assertNoConflictingSrcset(dir: NgOptimizedImage) {
+        `${
+            imgDirectiveDetails(
+                dir.ngSrc)} one or more breakpoints in the \`subViewportBreakpoints\` ` +
+            `array are greater than a breakpoint in the \`viewportBreakpoints\` array. The \`subViewportBreakpoints\` ` +
+            `array should only include values smaller than the viewport breakpoints.`);
```
For consistency with other error messages, let's include a recommendation for the next step:
```suggestion
            `array are greater than a breakpoint in the \`viewportBreakpoints\` array. ` +
            `To fix this, update the \`IMAGE_CONFIG\` token value and make sure that the ` +
            `\`subViewportBreakpoints\` array includes values smaller than the viewport breakpoints.`);
```

atcastle replied:
Removed this error case entirely since there's now only one breakpoint array.

AndrewKushnir reviewed (commented):
@atcastle thanks for creating this PR! 👍 Just left a few comments.

pkozlowski-opensource commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -1025,6 +1010,192 @@ describe('Image directive', () => {
+                'http://localhost/img?w=16 16w, http://localhost/img?w=48 48w, http://localhost/img?w=640 640w, http://localhost/img?w=1280 1280w, http://localhost/img?w=3840 3840w');
+      });
+
+      it('should disable automatic srcset generation if disableAutoSrcset is set in config provider',
```
+1 

I wouldn't mind if we drop `disableAutoSrcset` config on the DI level

pkozlowski-opensource commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -1025,6 +1010,192 @@ describe('Image directive', () => {
+      let imageLoader!: ImageLoader;
+
+      beforeEach(() => {
+        imageLoader = (config: ImageLoaderConfig) => {
```
as far as I can tell the `imageLoader` function here is stateless so I'm not sure if we need to re-create it in the `beforeEach` block - could it be a `const` in the `describe` block?

atcastle replied:
Good catch, done.

atcastle commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -223,6 +262,12 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
   @Input() ngSrcset!: string;
 
+  /**
+   * The base 'sizes' attribute passed through to the img element.
```
Fixed

atcastle commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +66,38 @@ const ASPECT_RATIO_TOLERANCE = .1;
 const OVERSIZED_IMAGE_TOLERANCE = 1000;
 
+/**
+ * A configuration object for the ngOptimizedImage directive. Contains:
```
Fixed

atcastle commented on aio/content/guide/image-directive.md:
```diff
@@ -89,24 +89,64 @@ You can typically fix this by adding `height: auto` or `width: auto` to your ima
+Viewport sizes: `[640, 750, 828, 1080, 1200, 1920, 2048, 3840]`
+Sub-Viewport sizes (only used if the sizes property is less than 100vw): `[16, 32, 48, 64, 96, 128, 256, 384]`
+
+If you would like to customize either or both of these breakpoint arrays, you can do so using the `IMAGE-CONFIG` provider:
```
Fixed

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +70,37 @@ const ASPECT_RATIO_TOLERANCE = .1;
+ * Injection token that configures the image optimized image functionality.
+ *
+ * @see `NgOptimizedImage`
+ * @publicApi
```
Remove the public API annotation?

atcastle replied:
done

kara commented on aio/content/guide/image-directive.md:
```diff
@@ -89,24 +89,64 @@ You can typically fix this by adding `height: auto` or `width: auto` to your ima
+
+</code-example>
+
+ Alternatively, you can set the `disableAutoSrcset` property to `true` on the IMAGE_CONFIG provider:
+
+<code-example format="typescript" language="typescript">
+providers: [
+  {
+    provide: IMAGE_CONFIG,
+    useValue: {
+      disableAutoSrcset: true
+    }
+  },
+],
+</code-example>
+
```
We agreed to remove this feature, no?

```suggestion

```

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -269,6 +314,18 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   * Disables automatic srcset generation for this image.
+   */
+  @Input()
+  set unoptimized(value: string|boolean|undefined) {
```
Hmm I like the idea of making it more specific, but I also want to make it clear that this pattern is not recommended. What about "unoptimizedSrcset"?

atcastle replied:
To me, `unoptimizedSrcset` sort of implies that you get a srcset, but that it's not optimized. But with this attribute you actually get nothing, srcset-wise.

kara replied:
Oh, that's a good point. I can see it being read that way 🤔. I'm good with "unoptimized"

AndrewKushnir replied:
I think I'm still a bit concerned that we may be delivering a wrong message via an input name. The `unoptimized` sounds more like the entire optimization logic (including loaders, etc) is disabled. Can we chat about this during the next sync?

It'd be ok to proceed with the PR as is and update it later (before the feature freeze).

kara commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -707,326 +689,481 @@ describe('Image directive', () => {
+                    'http://localhost/img?w=16 16w, http://localhost/img?w=48 48w, http://localhost/img?w=640 640w, http://localhost/img?w=1280 1280w, http://localhost/img?w=3840 3840w');
+          });
+
+          it('should disable automatic srcset generation if disableAutoSrcset is set in config provider',
```
Remove this test as feature is being removed?

atcastle replied:
Done

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +70,37 @@ const ASPECT_RATIO_TOLERANCE = .1;
 
+/**
+ * A configuration object for the NgOptimizedImage directive. Contains:
+ * - disableAutoSrcset: Turns off the automated srcset function.
```
```suggestion
```

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +70,37 @@ const ASPECT_RATIO_TOLERANCE = .1;
+ * @developerPreview
+ */
+export type ImageConfig = {
+  disableAutoSrcset?: boolean,
```
Feature should be removed

```suggestion
```

kara commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -707,326 +689,481 @@ describe('Image directive', () => {
+                       'http://localhost/img?w=16 16w, http://localhost/img?w=32 32w, http://localhost/img?w=48 48w, http://localhost/img?w=64 64w, http://localhost/img?w=96 96w, http://localhost/img?w=128 128w, http://localhost/img?w=256 256w, http://localhost/img?w=384 384w, http://localhost/img?w=640 640w, http://localhost/img?w=750 750w, http://localhost/img?w=828 828w, http://localhost/img?w ...
+             });
+
+          it('should trim the responsive srcset if sizes attribute exists and is less than 100vw',
```
Can we add tests for:

- sizes with a media query 
- sizes with a media query that needs an ngSrcset

atcastle replied:
Done

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -223,6 +263,12 @@ export class NgOptimizedImage implements OnInit, OnChanges, OnDestroy {
+   * The base `sizes` attribute passed through to the `<img>` element.
+   * Providing sizes causes the image to use an automatic responsive srcset.
+   */
+  @Input() sizes?: string;
```
You're going to need to set a host attribute for this (ex: L389) to make sure it's not swallowed by the input decorator (i.e. if you make it into an input, it's no longer an attribute in the DOM unless you manually copy it over). Can we also add a test that ensures "sizes" is on the actual HTML element?

atcastle replied:
done and done

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +70,34 @@ const ASPECT_RATIO_TOLERANCE = .1;
+ * @see `NgOptimizedImage`
+ * @developerPreview
+ */
+export const IMAGE_CONFIG = new InjectionToken<ImageConfig>(
```
Question: do we want people to provide different breakpoint values in the different parts of the application? Or is the intention that one set of breakpoints applies to the entire application?

atcastle replied:
I believe the main use case for doing this is to match a configuration on your CDN. Since it's possible to have different CDNs for different sections of an application (one for product images, another for static images, for instance) I think it makes sense to be able to set this differently for different parts.

atcastle committed:
feat(common): Add automatic srcset generation to ngOptimizedImage

Add a feature to automatically generate the srcset attribute for images using the NgOptimizedImage directive. Uses the 'sizes' attribute to determine the appropriate srcset to generate.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +70,34 @@ const ASPECT_RATIO_TOLERANCE = .1;
+ *      srcsets for responsive images.
+ * - subViewportBreakpoints: An array of smaller breakpoints. Added to the
+ *       srcset when `sizes` indicates a sub-full-width responsive image.
+ *
```
It looks like the docs are outdated (they refer to the previous version of the config).

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +70,34 @@ const ASPECT_RATIO_TOLERANCE = .1;
+ *       srcset when `sizes` indicates a sub-full-width responsive image.
+ *
+ * Learn more about the responsive image configuration in [the NgOptimizedImage
+ * guide](guide/image-directive).
+ * @developerPreview
```
```suggestion
 * guide](guide/image-directive).
 * @publicApi
 * @developerPreview
```

atcastle replied:
done

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -61,6 +70,34 @@ const ASPECT_RATIO_TOLERANCE = .1;
+/**
+ * Injection token that configures the image optimized image functionality.
+ *
+ * @see `NgOptimizedImage`
+ * @developerPreview
+ */
```
```suggestion
 * @see `NgOptimizedImage`
 * @publicApi
 * @developerPreview
 */
```

atcastle replied:
done

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -448,6 +555,23 @@ function assertNotBase64Image(dir: NgOptimizedImage) {
+ */
+function assertNoComplexSizes(dir: NgOptimizedImage) {
+  let sizes = dir.sizes;
+  if (sizes) {
+    if (sizes.match(/((\)|,)\s|^)\d+px/)) {
```
It looks like we can join 2 `if`s:
```suggestion
  if (sizes && sizes.match(/((\)|,)\s|^)\d+px/)) {
```

atcastle replied:
done

AndrewKushnir commented on packages/core/test/bundling/image-directive/e2e/image-distortion/image-distortion.ts:
```diff
@@ -60,18 +60,18 @@ export class ImageDistortionPassingComponent {
      <!-- Supplied aspect ratio differs from intrinsic aspect ratio by > .1 -->
      <!-- Aspect-ratio: 0.86666666666 -->
-     <img ngSrc="/e2e/b.png" width="26" height="30">
+     <img ngSrc="/e2e/b.png" width="26" height="30" unoptimized>
```
Just curious why we need to add `unoptimized` here?

AndrewKushnir commented:
[Presubmit](http://test/OCL:479665896:BASE:479665942:1665177141671:1381a004).

atcastle commented:
Latest update to this PR changes `unoptimized` attribute name to `disableOptimizedSrcset`

kara commented on packages/common/test/directives/ng_optimized_image_spec.ts:
```diff
@@ -1033,6 +1018,180 @@ describe('Image directive', () => {
+        setupTestingModule();
+
+        const template =
+            '<img ngSrc="path/img.png" width="100" height="50" sizes="(min-width: 768px) 500px, 100vw" srcset="www.example.com/img.png?w=500 768w, www.example.com/img.png?w=2000" >';
```
This should actually throw because `srcset` is being used. It has to be `ngSrcset`.

atcastle replied:
Got it. I got those two confused last minute and swapped them because I was thinking that ngSrcset can't handle complex srcsets, but that's not actually the case. Fixed

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -448,6 +555,21 @@ function assertNotBase64Image(dir: NgOptimizedImage) {
+        `${imgDirectiveDetails(dir.ngSrc, false)} \`sizes\` was set to a string including ` +
+            `pixel values. For automatic \`srcset\` generation, \`sizes\` must only include responsive ` +
+            `values, such as \`sizes="50vw"\` or \`sizes="(min-width: 768px) 50vw, 100vw"\`. ` +
+            `To fix this, modify the \`sizes\` attribute, or provide your own \`srcset\` value directly.`);
```
We shouldn't recommend people add `srcset` directly

```suggestion
            `To fix this, modify the \`sizes\` attribute, or provide your own \`ngSrcset\` value directly.`);
```

atcastle replied:
Done

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
Merge-assistance (@jessicajaniuk): this PR is almost ready for merge, needs one more `public-api` approval. Could you please take a look from the `public-api` perspective? If everything is ok, we can proceed with the merge. Thank you.

jessicajaniuk reviewed (approved):
reviewed-for: public-api

jessicajaniuk commented:
This PR was merged into the repository by commit 4fde292bb58f5d5bc3cf6e634f7cff9eb0d13d84.
